timemachine/cpp/src/tibd_exchange_move.cu

```diff
@@ -0,0 +1,317 @@
+
+    d_sort_storage_.realloc(sort_storage_bytes_);
+}
+
+template <typename RealType> TIBDExchangeMove<RealType>::~TIBDExchangeMove() {
+    gpuErrchk(cudaEventDestroy(host_copy_event_));
+}
+
+template <typename RealType>
+void TIBDExchangeMove<RealType>::move_device(
+    const int N,
+    double *d_coords, // [N, 3]
+    double *d_box,    // [3, 3]
+    cudaStream_t stream) {
+
+    if (N != this->N_) {
+        throw std::runtime_error("N != N_");
+    }
+
+    // Set the stream for the generator
+    curandErrchk(curandSetStream(this->cr_rng_, stream));
+
+    this->compute_initial_weights(N, d_coords, d_box, stream);
+
+    const int tpb = DEFAULT_THREADS_PER_BLOCK;
+    const int mol_blocks = ceil_divide(this->num_target_mols_, tpb);
+
+    dim3 atom_by_atom_grid(ceil_divide(N, tpb), this->mol_size_, 1);
+
+    k_compute_centroid_of_atoms<RealType><<<ceil_divide(d_ligand_idxs_.length, tpb), tpb, 0, stream>>>(
+        static_cast<int>(d_ligand_idxs_.length), d_ligand_idxs_.data, d_coords, d_center_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    k_compute_box_volume<<<1, 1, 0, stream>>>(d_box, d_box_volume_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    const int num_samples = NUM_SAMPLES;
+    for (int move = 0; move < this->proposals_per_move_; move++) {
+        // Run only after the first pass, to maintain meaningful `log_probability_host` values
+        if (move > 0) {
+            // Run a separate kernel to replace the before log probs and weights with the after if accepted a move
+            // Need the weights to sample a value and the log probs are just because they aren't expensive to copy
+            k_store_accepted_log_probability_targeted<RealType><<<1, tpb, 0>>>(
+                this->num_target_mols_,
+                d_targeting_inner_vol_.data,
+                d_box_volume_.data,
+                inner_volume_,
+                this->d_acceptance_.data + 1, // Offset to get the last value for the acceptance criteria
+                this->d_log_sum_exp_before_.data,
+                this->d_log_sum_exp_after_.data,
+                this->d_log_weights_before_.data,
+                this->d_log_weights_after_.data);
+            gpuErrchk(cudaPeekAtLastError());
+        }
+        // Copy the before log weights to the after weights, we will adjust the after weights incrementally
+        gpuErrchk(cudaMemcpyAsync(
+            this->d_log_weights_after_.data,
+            this->d_log_weights_before_.data,
+            this->d_log_weights_after_.size(),
+            cudaMemcpyDeviceToDevice,
+            stream));
+
+        gpuErrchk(cudaMemsetAsync(d_inner_mols_count_.data, 0, d_inner_mols_count_.size(), stream));
+        gpuErrchk(cudaMemsetAsync(d_outer_mols_count_.data, 0, d_outer_mols_count_.size(), stream));
+
+        k_split_mols_inner_outer<RealType><<<mol_blocks, tpb, 0, stream>>>(
+            this->num_target_mols_,
```

Comment:
If this is a bottleneck, we can probably cache this in an optimization pass as well (just like log weights, the assignment of water into different groups only needs to happen when a move is accepted).

Reply:
Right now the bottleneck is getting the number of waters in each region as well as whether or not you are targeting the center. 

To push everything to the GPU it seems like you would need to do the following:
  * Need a new kernel for computing argmax that doesn't use cub (requires a value for N) or set a min weight on all the 'empty values'
  * Set a dummy value for the sort
  * Handle the targeting mode in all of the individual kernels